### PR TITLE
refactor: Move `queueWorker` logic to CE & add it to omnichannel service

### PR DESCRIPTION
**Pull request by KevLehman**
OC-1109

### CODE REVIEW

File: apps/meteor/server/services/omnichannel/logger.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import { Logger } from '../../lib/logger/Logger';
+
+export const queueLogger = new Logger('OmnichannelQueue');
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel/logger.ts:3]: Could you cleanup this old logger? I don't think anyone is using it and since they both have the same name, it might create confusion

File: apps/meteor/server/services/omnichannel/queue.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+import { dispatchAgentDelegated } from '../../../app/livechat/server/lib/Helper';
+
+const DEFAULT_RACE_TIMEOUT = 5000;
+
+export class OmnichannelQueue implements IOmnichannelQueue {
+	private running = false;
+
+	private queues: (string | undefined)[] = [];
+
+	delay() {
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel/queue.ts:17]: This could be private func

File: apps/meteor/server/services/omnichannel/queue.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	async stop() {
+		queueLogger.debug('Stopping queue');
+		await LivechatInquiry.unlockAll();
+
+		this.running = false;
+	}
+
+	async getActiveQueues() {
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel/queue.ts:43]: This can also be private

File: apps/meteor/server/services/omnichannel/queue.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+		}
+
+		const queue = await this.nextQueue();
+		const queueDelayTimeout = this.delay();
+		queueLogger.debug(`Executing queue ${queue || 'Public'} with timeout of ${queueDelayTimeout}`);
+
+		setTimeout(this.checkQueue.bind(this, queue), queueDelayTimeout);
+	}
+
+	async checkQueue(queue: string | undefined) {
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel/queue.ts:70]: This could be private func

File: apps/meteor/server/services/omnichannel/queue.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+		queueLogger.debug({
+			msg: 'Routing method supports auto assignment',
+			method: RoutingManager.methodName,
+			status: routingSupportsAutoAssign ? 'Starting' : 'Stopping',
+		});
+
+		void (routingSupportsAutoAssign ? this.start() : this.stop());
+	}
+
+	async processWaitingQueue(department: string | undefined, inquiry: InquiryWithAgentInfo) {
```
Comment (murtaza98) [inline apps/meteor/server/services/omnichannel/queue.ts:114]: This could be private func

File: apps/meteor/server/services/omnichannel/logger.ts (hunk 1 above)
Comment (KevLehman) [reply apps/meteor/server/services/omnichannel/logger.ts:3]: This is the new logger lol

But i removed the old one :awesome:

File: packages/core-typings/src/omnichannel/queue.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+	delay(): number;
+	start(): Promise<void>;
+	shouldStart(): void;
+	stop(): Promise<void>;
+	getActiveQueues(): Promise<(string | undefined)[]>;
+	nextQueue(): Promise<string | undefined>;
+	execute(): Promise<void>;
+	checkQueue(queue: string | undefined): Promise<void>;
+	processWaitingQueue(department: string | undefined, inquiry: InquiryWithAgentInfo): Promise<boolean>;
+}
```
Comment (murtaza98) [inline packages/core-typings/src/omnichannel/queue.ts:13]:
```suggestion
export interface IOmnichannelQueue {
	start(): Promise<void>;
	shouldStart(): void;
	stop(): Promise<void>;
}
```

Comment (murtaza98) [reply packages/core-typings/src/omnichannel/queue.ts:13]: removing all private methods from public interface

**Review by murtaza98 — APPROVED** [review]

**Review by murtaza98 — DISMISSED** [review]

**Comment by sampaiodiego** [comment]
I noticed that the code in `develop` is creating a collection named `_queue` .. can we fix that, maybe on this PR?

File: apps/meteor/server/models/raw/LivechatInquiry.ts [hunk 7]
```diff
@@ -172,7 +172,10 @@ export class LivechatInquiryRaw extends BaseRaw<ILivechatInquiryRecord> implemen
 	}
 
 	async unlock(inquiryId: string): Promise<UpdateResult> {
-		return this.updateOne({ _id: inquiryId }, { $unset: { locked: 1, lockedAt: 1 } });
+		return this.updateOne(
+			{ _id: inquiryId },
+			{ $unset: { locked: 1, lockedAt: 1 }, $set: { status: LivechatInquiryStatus.QUEUED, queuedAt: new Date() } },
```
Comment (MarcosSpessatto) [inline apps/meteor/server/models/raw/LivechatInquiry.ts:177]: Wouldn't it be better if we create another PR with this change? It looks like it has been changing the behavior, isn't it?

**Comment by KevLehman** [comment]
@sampaiodiego the `_queue` collection is from the queue we use to process PDFs 👀

**Review by MarcosSpessatto — DISMISSED** [review]

**Review by murtaza98 — APPROVED** [review]

**Review by murtaza98 — APPROVED** [review]
